src/gaussian.cc

```diff
@@ -0,0 +1,185 @@
+/*
+ * This file is part of gauss2d.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef __GAUSS2D_GAUSSIAN_H_
+#include "gaussian.h"
+
+#include <string>
+#include <stdexcept>
+#include <cmath>
+
+namespace gauss2d
+{
+    void Covariance::init(double sigma_x, double sigma_y, double rho) {
+        Ellipse::check(sigma_x, sigma_y, rho);
+        this->_sigma_x_sq = sigma_x*sigma_x;
+        this->_sigma_y_sq = sigma_y*sigma_y;
+        this->_cov_xy = sigma_x * sigma_y * rho;
+    }
+
+    Covariance::Covariance(const Ellipse & ell) {
+        this->init(ell.get_sigma_x(), ell.get_sigma_y(), ell.get_rho());
+    }
+
+    std::shared_ptr<EllipseMajor> Covariance::make_ellipse_major(bool degrees) const
+    {
+        if(_sigma_x_sq == 0 && _sigma_y_sq == 0) return std::make_shared<EllipseMajor>(0, 1, 0);
+        double apc = _sigma_x_sq + _sigma_y_sq;
+        double x = apc/2;
+        double pm = sqrt(apc*apc - 4*(_sigma_x_sq*_sigma_y_sq - _cov_xy*_cov_xy)) / 2;
+
+        double r_major = x + pm;
+        if(r_major == 0) return std::make_shared<EllipseMajor>(0, 1, 0);
+        double axrat = sqrt((x - pm)/r_major);
+        r_major = sqrt(r_major);
```

Comment:
You should define `r_major = np.sqrt(x + pm)` initially and keep it out of the `sqrt` instead of redefining `r_major` here. The initial definition seems to serve no big purpose apart from defining `axrat`.